### PR TITLE
Conf Admin

### PR DESCRIPTION
**Pull request by mzurbuchen**

### CODE REVIEW

**Comment by psunix** [comment]
- new root pw task can be executed by Conf Admin -> but it shouldn't (task will be removed anyways)
- user role drop-down looks a bit strange. make it more bootstrap like: https://www.w3schools.com/bootstrap/bootstrap_dropdowns.asp

File: app/controllers/api/admin/users_controller.rb [hunk 1]
```diff
@@ -7,17 +7,20 @@
 
 class Api::Admin::UsersController < Api::Admin::AdminController
 
-  def toggle_admin
+  def update_role
```
Comment (psunix) [inline app/controllers/api/admin/users_controller.rb]: move to separate controller: api/admin/users/role_controller.rb
action: #update 
http://jeromedalbert.com/how-dhh-organizes-his-rails-controllers/

File: app/controllers/teams_controller.rb [hunk 2]
```diff
@@ -72,23 +76,14 @@ def team_params
...
-    flash[:error] = t('flashes.teams.cannot_delete')
-    redirect_to teams_path
+  def team
+    @team ||= Team.find(params[:id])
   end
 
-  def can_delete_team?(_team)
-    current_user.admin?
-  end
+  def user_not_authorized(exception)
```
Comment (psunix) [inline app/controllers/teams_controller.rb]: why don't we use the ApplicationController#user_not_authorized here ?
maybe leave a comment or remove this

File: app/helpers/admin/users_helper.rb [hunk 3]
```diff
@@ -16,4 +16,9 @@ def change_admin_attribute(user)
       content_tag(:button)
     end
   end
+
+  def role(userrole)
```
Comment (psunix) [inline app/helpers/admin/users_helper.rb]: beware! helper uses global namespace. so the method role is defined everywhere in views. 
rename #role

File: app/helpers/admin/users_helper.rb [hunk 4]
```diff
@@ -16,4 +16,9 @@ def change_admin_attribute(user)
...
+    roles = ['User', 'Conf Admin', 'Admin']
+    roles[User.roles[userrole]]
+  end
```
Comment (psunix) [inline app/helpers/admin/users_helper.rb]: too much logic inside view. move as much as possible to view helper -> app/views/users/index.html.haml
try to create whole dropdown with a helper

File: app/models/user.rb [hunk 5]
```diff
@@ -46,14 +46,17 @@ class User < ApplicationRecord
   scope :locked, (-> { where(locked: true) })
   scope :unlocked, (-> { where(locked: false) })
 
-  scope :admins, (-> { where(admin: true) })
+  scope :admins, (-> { where(role: :admin) })
+  scope :conf_admins, (-> { where(role: :conf_admin) })
```
Comment (psunix) [inline app/models/user.rb]: this scope seams not to be used anywhere ... remove it if so !

File: app/policies/account_policy.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
+class AccountPolicy < TeamDependantPolicy
+  def show?
```
Comment (psunix) [inline app/policies/account_policy.rb]: can't we move this methods also do TeamDepPol ? probably mostly the same for group, account, item

File: app/policies/team_dependant_policy.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
+class TeamDependantPolicy < ApplicationPolicy
+  def team_member?
+    team.teammember?(@user.id)
+  end
+
+  protected
+
+  def team
+    raise 'missing method'
```
Comment (psunix) [inline app/policies/team_dependant_policy.rb]: better: 'implement in subclass'

File: app/views/admin/users/index.html.haml [hunk 8]
```diff
@@ -23,17 +23,27 @@
           = user.last_login_at.strftime('%d.%m.%Y, %H:%M')
       %td= user.last_login_from
       %td= user.ldap_uid.to_s
+
       %td
         - if user != current_user
-          = change_admin_attribute(user)
+          .btn-group{id: user.id } 
+            %button.btn.btn-default.dropdown-toggle{id: 'role-dropdown','data-toggle':'dropdown'}
```
Comment (psunix) [inline app/views/admin/users/index.html.haml]: move to helper as mentioned before

Comment (psunix) [reply app/views/admin/users/index.html.haml]: http://api.rubyonrails.org/classes/ActionView/Helpers/TagHelper.html#method-i-content_tag

File: app/views/layouts/_navigation.html.haml [hunk 9]
```diff
@@ -11,15 +11,16 @@
       .navbar-collapse.collapse
         %ul.nav.navbar-nav
           = nav_link t('teams.title'), teams_path
-          - if current_user.admin?
+          - if current_user.admin? || current_user.conf_admin?
```
Comment (psunix) [inline app/views/layouts/_navigation.html.haml]: check by policy (pundit)

File: app/views/layouts/_navigation.html.haml [hunk 10]
```diff
@@ -11,15 +11,16 @@
...
             %li.dropdown
               <a href="#" data-target="#" class="dropdown-toggle" data-toggle="dropdown", role="button">
               %span=t('admin.title')
               %b.caret
               </a>
               %ul.dropdown-menu
                 %li= nav_link t(:settings), admin_settings_path
-                %li= nav_link t(:recrypt_requests), admin_recryptrequests_path
+                - if current_user.admin?
```
Comment (psunix) [inline app/views/layouts/_navigation.html.haml]: check by policy (pundit)

File: db/seeds/support/user_seeder.rb [hunk 11]
```diff
@@ -13,31 +13,35 @@ def seed_root
       u.username = 'root'
       u.givenname = 'root'
       u.auth = 'db'
-      u.admin = true
+      u.role = User::Role::ADMIN
```
Comment (psunix) [inline db/seeds/support/user_seeder.rb]: shouldn't this be u.role = :admin ?

Comment (psunix) [reply db/seeds/support/user_seeder.rb]: also check other User::Role ...

File: test/controllers/api/admin/users_controller_test.rb [hunk 12]
```diff
@@ -11,63 +11,313 @@ class Api::Admin::UsersControllerTest < ActionController::TestCase
 
   include ControllerTest::DefaultHelper
 
-  test 'admin empowers user' do
-    teammembers(:team1_bob).destroy
-    bob = users(:bob)
+  context '#update_role' do
+    context 'root' do
+      test 'root updates admin to user' do
+        teammembers(:team1_bob).destroy
```
Comment (psunix) [inline test/controllers/api/admin/users_controller_test.rb]: prefer destroy! unless you don't want to get noticed when it fails

File: test/controllers/api/admin/users_controller_test.rb [hunk 13]
```diff
@@ -11,63 +11,313 @@ class Api::Admin::UsersControllerTest < ActionController::TestCase
...
+
+      test 'conf admin cannot update user to admin' do
+        teammembers(:team1_bob).destroy
+        bob = users(:bob)
+
+        login_as(:tux)
+        patch :update_role, params: { user_id: bob, role: :admin }, xhr: true
+
+        bob.reload
+        assert bob.user?
```
Comment (psunix) [inline test/controllers/api/admin/users_controller_test.rb]: use assert_equals true, bob.user?
or check api doc for assert

File: test/controllers/api/admin/users_controller_test.rb [hunk 14]
```diff
@@ -11,63 +11,313 @@ class Api::Admin::UsersControllerTest < ActionController::TestCase
...
+        assert conf_admin.admin?
+        assert conf_admin.teammembers.find_by(team_id: teams(:team1))
+      end
+
+      test 'admin updates admin to conf_admin' do
+        teammembers(:team1_bob).destroy
+        admin2 = Fabricate(:admin)
+
+        login_as(:admin)
+        patch :update_role, params: { user_id: admin2, role: :conf_admin }, xhr: true
```
Comment (psunix) [inline test/controllers/api/admin/users_controller_test.rb]: move these tests to new controller -> see recent comment

File: test/policies/policy_test.rb [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+
+class PolicyTest < ActiveSupport::TestCase
+  protected
+  
+  def assert_permit(user, record, action)
+    msg = "User #{user.inspect} should be permitted to #{action} #{record}, but isn't permitted"
+    assert Pundit.policy!(user, record).public_send(action), msg
+  end
+
+  def refute_permit(user, record, action)
```
Comment (psunix) [inline test/policies/policy_test.rb]: nice! I like these helper actions ! well done !

File: test/controllers/api/admin/users_controller_test.rb (hunk 13 above)
Comment (mzurbuchen) [reply test/controllers/api/admin/users_controller_test.rb]: Done

File: app/controllers/admin/recryptrequests_controller.rb [hunk 16]
```diff
@@ -46,6 +46,12 @@ def resetpassword
 
   private
 
+  def authorize_and_set_user_and_admin
```
[the remark this answers is not in the document]
Comment (DominikFischli) [reply app/controllers/admin/recryptrequests_controller.rb]: Done

File: test/controllers/api/admin/users_controller_test.rb (hunk 12 above)
Comment (mzurbuchen) [reply test/controllers/api/admin/users_controller_test.rb]: Done

File: db/seeds/support/user_seeder.rb (hunk 11 above)
Comment (mzurbuchen) [reply db/seeds/support/user_seeder.rb]: Done

File: app/policies/team_dependant_policy.rb (hunk 7 above)
Comment (mzurbuchen) [reply app/policies/team_dependant_policy.rb]: Done

File: app/policies/account_policy.rb (hunk 6 above)
Comment (mzurbuchen) [reply app/policies/account_policy.rb]: Done

Comment (mzurbuchen) [reply app/policies/account_policy.rb]: Done

File: app/models/user.rb (hunk 5 above)
Comment (mzurbuchen) [reply app/models/user.rb]: Done

File: app/views/layouts/_navigation.html.haml (hunk 9 above)
Comment (mzurbuchen) [reply app/views/layouts/_navigation.html.haml]: Done

File: app/views/layouts/_navigation.html.haml (hunk 10 above)
Comment (mzurbuchen) [reply app/views/layouts/_navigation.html.haml]: Done

File: app/helpers/admin/users_helper.rb (hunk 3 above)
Comment (mzurbuchen) [reply app/helpers/admin/users_helper.rb]: Done

File: app/controllers/api/admin/users_controller.rb (hunk 1 above)
Comment (DominikFischli) [reply app/controllers/api/admin/users_controller.rb]: Done

File: test/policies/team_policy_test.rb [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+      Team.all.each do |t|
+        unless t.private? do assert teams.include?(t); end
+        end
+      end   
+    end
+
+    test 'user can see all and only the teams he is a part of' do
+      teams = TeamPolicy::Scope.new(bob, Team).resolve
+      bob.teams.each do |t|
+        assert teams.include?(t)
```
Comment (psunix) [inline test/policies/team_policy_test.rb]: don't use assert ... assert_equals

File: test/policies/team_policy_test.rb [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+    test 'conf admin cannot delete a team with more than one teammmember' do
+      refute_permit conf_admin, team1, :destroy?
+    end
+  end
+
+  context '#scope' do
+    test 'admin sees all non-private teams' do
+      teams = TeamPolicy::Scope.new(admin, Team).resolve
+      Team.all.each do |t|
+        unless t.private? do assert teams.include?(t); end
```
Comment (psunix) [inline test/policies/team_policy_test.rb]: ugly !!!!
better use Team.where or scope for private teams

File: test/policies/team_policy_test.rb [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+
+  def team1
+    teams(:team1)
+  end
+
+  def team2
+    teams(:team2)
+  end
+
+  def private_team
```
Comment (psunix) [inline test/policies/team_policy_test.rb]: use fabricator !!!

**Comment by psunix** [comment]
done in other pull request

**Pull request closed without merging**